// element-plus/lib/components/config-provider/src/config-provider-props.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_runtime = require("../../../utils/vue/props/runtime.js");
const require_index = require("../../../hooks/use-size/index.js");
const require_index$1 = require("../../../hooks/use-empty-values/index.js");
//#region ../../packages/components/config-provider/src/config-provider-props.ts
const configProviderProps = require_runtime.buildProps({
	/**
	* @description Controlling if the users want a11y features
	*/
	a11y: {
		type: Boolean,
		default: true
	},
	/**
	* @description Locale Object
	*/
	locale: { type: require_runtime.definePropType(Object) },
	/**
	* @description global component size
	*/
	size: require_index.useSizeProp,
	/**
	* @description button related configuration, [see the following table](https://element-plus.org/en-US/component/config-provider.html#button-attribute)
	*/
	button: { type: require_runtime.definePropType(Object) },
	/**
	* @description card related configuration, [see the following table](https://element-plus.org/en-US/component/config-provider.html#card-attribute)
	*/
	card: { type: require_runtime.definePropType(Object) },
	/**
	* @description dialog related configuration, [see the following table](https://element-plus.org/en-US/component/config-provider.html#dialog-attribute)
	*/
	dialog: { type: require_runtime.definePropType(Object) },
	/**
	* @description link related configuration, [see the following table](https://element-plus.org/en-US/component/config-provider.html#link-attribute)
	*/
	link: { type: require_runtime.definePropType(Object) },
	/**
	* @description features at experimental stage to be added, all features are default to be set to false, [see the following table](https://element-plus.org/en-US/component/config-provider.html#experimental-features)                                                                            | ^[object]
	*/
	experimentalFeatures: { type: require_runtime.definePropType(Object) },
	/**
	* @description Controls if we should handle keyboard navigation
	*/
	keyboardNavigation: {
		type: Boolean,
		default: true
	},
	/**
	* @description message related configuration, [see the following table](https://element-plus.org/en-US/component/config-provider.html#message-attribute)
	*/
	message: { type: require_runtime.definePropType(Object) },
	/**
	* @description global Initial zIndex
	*/
	zIndex: Number,
	/**
	* @description global component className prefix (cooperated with [$namespace](https://github.com/element-plus/element-plus/blob/dev/packages/theme-chalk/src/mixins/config.scss#L1)) | ^[string]
	*/
	namespace: {
		type: String,
		default: "el"
	},
	/**
	* @description table related configuration, [see the following table](https://element-plus.org/en-US/component/config-provider.html#table-attribute)
	*/
	table: { type: require_runtime.definePropType(Object) },
	...require_index$1.useEmptyValuesProps
});
//#endregion
exports.configProviderProps = configProviderProps;
